### PR TITLE
Maintain a RESTMapper per cluster to avoid performing a discovery on each API call

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
**tl;dr**
The dynamic client performs 33 API calls during creation - Bad for the Kubermatic API latency (When loading Nodes). This PR avoids those calls.

Everytime we create a dynamic client, it performs a discovery process.
The discovery process looks at all available API groups on a Kubernetes API server & fetches the types for each group. This results in a lot if API calls:
```
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/api?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/api/v1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/apiregistration.k8s.io/v1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/apiregistration.k8s.io/v1beta1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/extensions/v1beta1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/apps/v1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/apps/v1beta2?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/apps/v1beta1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/events.k8s.io/v1beta1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/authentication.k8s.io/v1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/authentication.k8s.io/v1beta1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/authorization.k8s.io/v1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/authorization.k8s.io/v1beta1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/autoscaling/v1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/autoscaling/v2beta1?timeout=32s
2019/05/02 17:49:11 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/autoscaling/v2beta2?timeout=32s
2019/05/02 17:49:12 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/batch/v1?timeout=32s
2019/05/02 17:49:12 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/batch/v1beta1?timeout=32s
2019/05/02 17:49:12 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/certificates.k8s.io/v1beta1?timeout=32s
2019/05/02 17:49:12 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/networking.k8s.io/v1?timeout=32s
2019/05/02 17:49:12 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/policy/v1beta1?timeout=32s
2019/05/02 17:49:12 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/rbac.authorization.k8s.io/v1?timeout=32s
2019/05/02 17:49:13 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/rbac.authorization.k8s.io/v1beta1?timeout=32s
2019/05/02 17:49:13 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/storage.k8s.io/v1?timeout=32s
2019/05/02 17:49:13 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/storage.k8s.io/v1beta1?timeout=32s
2019/05/02 17:49:13 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/admissionregistration.k8s.io/v1beta1?timeout=32s
2019/05/02 17:49:13 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/apiextensions.k8s.io/v1beta1?timeout=32s
2019/05/02 17:49:13 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/scheduling.k8s.io/v1beta1?timeout=32s
2019/05/02 17:49:13 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/coordination.k8s.io/v1beta1?timeout=32s
2019/05/02 17:49:13 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/crd.projectcalico.org/v1?timeout=32s
2019/05/02 17:49:13 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/cluster.k8s.io/v1alpha1?timeout=32s
2019/05/02 17:49:13 https://wflspwnjkn.europe-west3-c.dev.kubermatic.io:32640/apis/metrics.k8s.io/v1beta1?timeout=32s
```

The Kubermatic API unfortunately creates a new dynamic client during each API call which needs to fetch data from a user cluster.
When a user opens the cluster detail page, the Dashboard performs 2 API calls which fetch data from the user cluster:
```
127.0.0.1 - - [02/May/2019:17:58:19 +0200] "GET /api/v1/projects/gjzf628nsg/dc/europe-west3-c/clusters/wflspwnjkn/nodedeployments HTTP/1.1" 200 517 "http://localhost:8000/projects/gjzf628nsg/dc/europe-west3-c/clusters/wflspwnjkn" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.108 Safari/537.36"
127.0.0.1 - - [02/May/2019:17:58:19 +0200] "GET /api/v1/projects/gjzf628nsg/dc/europe-west3-c/clusters/wflspwnjkn/nodes?hideInitialConditions=true HTTP/1.1" 200 2 "http://localhost:8000/projects/gjzf628nsg/dc/europe-west3-c/clusters/wflspwnjkn" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.108 Safari/537.36"
```

This leads to 2 discoveries, each triggering 33 API calls (For an empty user cluster).

To avoid this, i'm introducing a map of "RESTMappers" - For each cluster 1.

But instead if using the default RESTMapper, i'm introducing a wrapper around it.
This avoids having an outdated mapping, which causes errors.
Example:
If the normal RESTMapper got created before we created the Machine/MachineSet/MachineDeployment CRD's, the RESTMapper would not be able to lookup the type information for those types & it will not try to load them. The default RESTMapper only loads the infos during creation.
Upstream issue: https://github.com/kubernetes-sigs/controller-runtime/issues/321

The introduced wrapper will invalidate the mapping and retry the lookup once when it hits an error.

**Does this PR introduce a user-facing change?**:
```release-note
Reduced API latency for loading Nodes & NodeDeployments
```

### CODE REVIEW

**Comment by mrIncompetent** [comment]
cc @themue

File: api/pkg/cluster/client/client.go [hunk 1]
```diff
@@ -110,11 +152,16 @@ func (p *provider) GetClientConfig(c *kubermaticv1.Cluster, options ...ConfigOpt
 
 // GetClient returns a dynamic client
 func (p *provider) GetClient(c *kubermaticv1.Cluster, options ...ConfigOption) (ctrlruntimeclient.Client, error) {
+	mapper, err := p.getMapper(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the REST mapper for the client: %v", err)
+	}
 	config, err := p.GetClientConfig(c, options...)
```
Comment (alvaroaleman) [inline api/pkg/cluster/client/client.go]: Can we please just pass in and use this config for `getMapper`? This also spares us one call to `GetConfig` if the mapper is not there

File: api/pkg/util/restmapper/invalidate.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	return m.lastError
+}
+
+func (m *invalidationMapper) KindFor(resource schema.GroupVersionResource) (gvk schema.GroupVersionKind, err error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	for i := 1; i <= 2; i++ {
```
Comment (alvaroaleman) [inline api/pkg/util/restmapper/invalidate.go]: This feels a bit overly complicated, what about doing something like:

```
import "k8s.io/apimachinery/pkg/api/meta"
if gvk, err = m.mapper.KindFor(resource); err != nil {
  if _, isNoKindMatchErr := err.(*meta.NoKindMatchError); isNoKindMatchErr {
    m.mapper, err = m.newRESTMapping()
    if err != nil {
      return err
    }
    return m.mapper.KindFor(resource)
  }
  return err
}
```

Used https://github.com/openshift/cluster-network-operator/pull/95/files as inspiration

**Review by alvaroaleman — COMMENTED** [review]
Two nits, but seems like a great change!

File: api/pkg/util/restmapper/invalidate.go (hunk 2 above)
Comment (alvaroaleman) [reply api/pkg/util/restmapper/invalidate.go]: The whole checking for the `NoKindMatchError` and if that is the case updating the mapper should probably moved into a `func (m *invalidationMapper) refreshMapperIfIsNoKindMatchErr(err error) error`

Comment (mrIncompetent) [reply api/pkg/util/restmapper/invalidate.go]: Haven't seen the openshift code - i'll copy it

File: api/pkg/cluster/client/client.go (hunk 1 above)
Comment (mrIncompetent) [reply api/pkg/cluster/client/client.go]: Done with https://github.com/kubermatic/kubermatic/pull/3405/commits/1671abad1ea36fcab955c2fed05c9772a7d431c8

File: api/pkg/util/restmapper/invalidate.go (hunk 2 above)
Comment (mrIncompetent) [reply api/pkg/util/restmapper/invalidate.go]: Done with https://github.com/kubermatic/kubermatic/pull/3405/commits/1671abad1ea36fcab955c2fed05c9772a7d431c8

File: api/pkg/util/restmapper/dynamic.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	if err != nil {
+		return err
+	}
+	drm.delegate = restmapper.NewDiscoveryRESTMapper(gr)
+	return nil
+}
+
+// reloadOnError checks if an error indicates that the delegated RESTMapper needs to be
+// reloaded, and if so, reloads it and returns true.
+func (drm *DynamicRESTMapper) reloadOnError(err error) bool {
```
Comment (alvaroaleman) [inline api/pkg/util/restmapper/dynamic.go:56]: This func is a bit weird IMHO, can we change it to

`func (drm *DynamicRESTMapper) reloadOnError(err error) (bool, error)`

?

Comment (mrIncompetent) [reply api/pkg/util/restmapper/dynamic.go:56]: I would leave it as it is - changing the signature as you provided would mean we would now need to introduce a dedicated error handling in each consumer.

Going the way via `utilruntime.HandleError(err)` is a nice compromise here as we already have an error which we would return.

File: api/pkg/cluster/client/client.go [hunk 4]
```diff
@@ -40,12 +50,39 @@ func NewExternal(seedClient ctrlruntimeclient.Client) (UserClusterConnectionProv
...
+		mapperLock:         &sync.Mutex{},
+		clusterRESTMapper:  map[string]meta.RESTMapper{},
+	}, nil
 }
 
 type provider struct {
 	seedClient         ctrlruntimeclient.Client
 	useExternalAddress bool
+
+	mapperLock *sync.Mutex
```
Comment (themue) [inline api/pkg/cluster/client/client.go]: No need to make it a reference and explicitly create a mutex. Typical pattern is 

```go
type provider struct {
    mu sync.Mutex
    ...
}

func (p *provider) mapper(...) (meta.RESTMapper, error) {
    p.mu.Lock()
    defer p.mu.Unlock()

    ...
}
```

See e.g. source code of HTTP server.

**Comment by alvaroaleman** [comment]
/lgtm
/approve
/hold

so @p0lyn0mial can have a look

File: api/pkg/cluster/client/client.go [hunk 5]
```diff
@@ -40,12 +50,39 @@ func NewExternal(seedClient ctrlruntimeclient.Client) (UserClusterConnectionProv
...
+	// We keep the existing cluster mappings to avoid the discovery on each call to the API server
+	clusterRESTMapper map[string]meta.RESTMapper
+}
+
+func (p *provider) getMapper(c *kubermaticv1.Cluster, config *restclient.Config) (meta.RESTMapper, error) {
```
Comment (themue) [inline api/pkg/cluster/client/client.go]: Are we forced to use get prefixes here and in general? It's not convention, see https://golang.org/doc/effective_go.html#Getters . Sadly it finds its way into Go code by devs coming from other languages. Seen it a lot in Ethereum code too.

File: api/pkg/util/restmapper/dynamic.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+		return false
+	}
+	err = drm.reload()
+	if err != nil {
+		utilruntime.HandleError(err)
+	}
+	return err == nil
+}
+
+func (drm *DynamicRESTMapper) KindFor(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
```
Comment (themue) [inline api/pkg/util/restmapper/dynamic.go:76]: Any reason why public methods of a public type have no godoc comment?

File: api/pkg/util/restmapper/dynamic.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	gvk, err := drm.delegate.KindFor(resource)
+	if drm.reloadOnError(err) {
+		gvk, err = drm.delegate.KindFor(resource)
+	}
+	return gvk, err
+}
+
+func (drm *DynamicRESTMapper) KindsFor(resource schema.GroupVersionResource) ([]schema.GroupVersionKind, error) {
+	gvks, err := drm.delegate.KindsFor(resource)
```
Comment (themue) [inline api/pkg/util/restmapper/dynamic.go]: All these repeating patterns look more like a retry pattern to me:

```go
func (drm *DynamicRESTMapper) KindsFor(resource schema.GroupVersionResource) (gvks []schema.GroupVersionKind, err error) {
    drm.retry(func() error {
        gvks, err = drm.delegate.KindsFor(resource)
        return err
    }, 2)
}
```

where `retry()` reloads as often as the error type is matching and the counter (here 2) is not reached.

**Review by themue — COMMENTED** [review]
Some comments.

File: api/pkg/cluster/client/client.go [hunk 8]
```diff
@@ -40,12 +50,39 @@ func NewExternal(seedClient ctrlruntimeclient.Client) (UserClusterConnectionProv
...
+
+	if mapper, found := p.clusterRESTMapper[c.Name]; found {
+		return mapper, nil
+	}
+
+	mapper, err := restmapper.NewDynamicRESTMapper(config)
+	if err != nil {
+		return nil, err
+	}
+	glog.Infof("Created a mapper for cluster %s", c.Name)
```
Comment (p0lyn0mial) [inline api/pkg/cluster/client/client.go]: isn't this a debug msg ?

File: api/pkg/cluster/client/client.go [hunk 9]
```diff
@@ -114,7 +151,13 @@ func (p *provider) GetClient(c *kubermaticv1.Cluster, options ...ConfigOption) (
 	if err != nil {
 		return nil, err
 	}
-	dynamicClient, err := ctrlruntimeclient.New(config, ctrlruntimeclient.Options{})
+
+	mapper, err := p.getMapper(c, config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the REST mapper for the client: %v", err)
```
Comment (p0lyn0mial) [inline api/pkg/cluster/client/client.go]: instead of failing we could create the `ctrlruntimeclient` as before without the `mapper`, the request will likely succeed. If we would like to know how many times we failed to create the mapper we could collect some metric.

File: api/pkg/util/restmapper/dynamic.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	return drm, nil
+}
+
+func (drm *DynamicRESTMapper) reload() error {
+	gr, err := restmapper.GetAPIGroupResources(drm.client)
+	if err != nil {
+		return err
+	}
+	drm.delegate = restmapper.NewDiscoveryRESTMapper(gr)
```
Comment (p0lyn0mial) [inline api/pkg/util/restmapper/dynamic.go]: this is a data race, `meta.RESTMapper` is created once and used by many threads. For example, one thread can call `KindFor` and reach this line (WRITE) and at the same time other thread calling `ResourceFor` can `drm.delegate.ResourceFor(input)` (READ).

I think we could use `sync.RWMutex` to guard the delegate variable.

File: api/pkg/util/restmapper/dynamic.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+type DynamicRESTMapper struct {
+	client   discovery.DiscoveryInterface
+	delegate meta.RESTMapper
+}
+
+// NewDynamicRESTMapper returns a RESTMapper that dynamically discovers resource
+// types at runtime. This is in contrast to controller-manager's default RESTMapper, which
+// only checks resource types at startup, and so can't handle the case of first creating a
+// CRD and then creating an instance of that CRD.
+func NewDynamicRESTMapper(cfg *rest.Config) (meta.RESTMapper, error) {
```
Comment (p0lyn0mial) [inline api/pkg/util/restmapper/dynamic.go:26]: since we use Impersonation and cache `meta.RESTMapper` we can create the mapper for one user and then reuse for other users. For example, imagine that we store the mapper for `Bob` then for every request that `John` makes the kubeAPI server sees it as if it was `Bob`. 

This is only inconvenient for auditing. We could change this by:
 - creating the mapper for anonymous user, discovery is available for unauthenticated users.
 - always create the mapper as an admin.

**Review by p0lyn0mial — CHANGES_REQUESTED** [review]

File: api/pkg/util/restmapper/dynamic.go (hunk 10 above)
Comment (p0lyn0mial) [reply api/pkg/util/restmapper/dynamic.go]: /cc @zreigz it would be really nice to run the API server for E2E tests with `-race` flag enabled.

File: api/pkg/util/restmapper/dynamic.go (hunk 11 above)
Comment (alvaroaleman) [reply api/pkg/util/restmapper/dynamic.go:26]: > creating the mapper for anonymous user, discovery is available for unauthenticated users.

This will change for 1.15: https://github.com/kubernetes/kubernetes/pull/73807

Comment (mrIncompetent) [reply api/pkg/util/restmapper/dynamic.go:26]: We actually reuse the mapping across the different users.
We use the cluster name as key for the map of mappers.

This means that unauthenticated users could have access to the discovery information, given that the kubermatic API actually exposes it.

I don't see any big downside of that though.

File: api/pkg/util/restmapper/dynamic.go (hunk 10 above)
Comment (mrIncompetent) [reply api/pkg/util/restmapper/dynamic.go]: Done with https://github.com/kubermatic/kubermatic/pull/3405/commits/4894d7a5279b66bc6c1a701a95b7bf7c8df65a5a

File: api/pkg/cluster/client/client.go (hunk 9 above)
Comment (mrIncompetent) [reply api/pkg/cluster/client/client.go]: Done with https://github.com/kubermatic/kubermatic/pull/3405/commits/4894d7a5279b66bc6c1a701a95b7bf7c8df65a5a

File: api/pkg/cluster/client/client.go (hunk 8 above)
Comment (mrIncompetent) [reply api/pkg/cluster/client/client.go]: Done with https://github.com/kubermatic/kubermatic/pull/3405/commits/4894d7a5279b66bc6c1a701a95b7bf7c8df65a5a

File: api/pkg/cluster/client/client.go (hunk 5 above)
Comment (mrIncompetent) [reply api/pkg/cluster/client/client.go]: Done with https://github.com/kubermatic/kubermatic/pull/3405/commits/4894d7a5279b66bc6c1a701a95b7bf7c8df65a5a

File: api/pkg/util/restmapper/dynamic.go (hunk 7 above)
Comment (mrIncompetent) [reply api/pkg/util/restmapper/dynamic.go]: I thought about it, but we explicitly want to only retry once & perfom a certain action if the first attempt fails. Which is why i would prefer the simplicity of the simple `if` for now

File: api/pkg/cluster/client/client.go (hunk 4 above)
Comment (mrIncompetent) [reply api/pkg/cluster/client/client.go]: Done with https://github.com/kubermatic/kubermatic/pull/3405/commits/4894d7a5279b66bc6c1a701a95b7bf7c8df65a5a

File: api/pkg/util/restmapper/dynamic.go (hunk 6 above)
Comment (mrIncompetent) [reply api/pkg/util/restmapper/dynamic.go:76]: Added with https://github.com/kubermatic/kubermatic/pull/3405/commits/522b1d4b8784e8d774653105231fb9da75a287a9

**Comment by mrIncompetent** [comment]
@alvaroaleman @p0lyn0mial @themue PTAL

File: api/pkg/util/restmapper/dynamic.go [hunk 12]
```diff
@@ -37,7 +41,13 @@ func (drm *DynamicRESTMapper) reload() error {
 	if err != nil {
 		return err
 	}
-	drm.delegate = restmapper.NewDiscoveryRESTMapper(gr)
+
+	// The discovery takes some time
```
Comment (p0lyn0mial) [inline api/pkg/util/restmapper/dynamic.go:45]: nit: not sure if this comment is informative :P

File: api/pkg/util/restmapper/dynamic.go [hunk 13]
```diff
@@ -54,58 +64,65 @@ func (drm *DynamicRESTMapper) reloadOnError(err error) bool {
...
+func (drm *DynamicRESTMapper) mapper() meta.RESTMapper {
+	drm.lock.RLock()
+	defer drm.lock.RUnlock()
+
+	return drm.delegate
+}
+
 func (drm *DynamicRESTMapper) KindFor(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
-	gvk, err := drm.delegate.KindFor(resource)
+	gvk, err := drm.mapper().KindFor(resource)
```
Comment (p0lyn0mial) [inline api/pkg/util/restmapper/dynamic.go:77]: took me some time to convince myself this is safe but it is - nice !

File: api/pkg/cluster/client/client.go [hunk 14]
```diff
@@ -40,12 +49,38 @@ func NewExternal(seedClient ctrlruntimeclient.Client) (UserClusterConnectionProv
...
+
+	if mapper, found := p.clusterRESTMapper[c.Name]; found {
+		return mapper, nil
+	}
+
+	mapper, err := restmapper.NewDynamicRESTMapper(config)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(4).Infof("Created a mapper for cluster %s", c.Name)
```
Comment (p0lyn0mial) [inline api/pkg/cluster/client/client.go:80]: no structured logging yet ?

Comment (mrIncompetent) [reply api/pkg/cluster/client/client.go:80]: Not yet :/ I'm working on it

Comment (p0lyn0mial) [reply api/pkg/cluster/client/client.go:80]: oh, the pr not merged yet https://github.com/kubermatic/kubermatic/pull/3296

**Comment by p0lyn0mial** [comment]
nice pull @mrIncompetent 

/lgtm
/approve
